aya-bpf/helpers: add documentation and implement more bpf_probe_read_* wrappers

willfindlay opened the pull request:
This patch adds some documentation to aya-bpf/helpers and adds documentation
for the module itself and for all of the wrappers currently defined in the module.

It also implements the rest of the bpf_probe_read_* wrappers that were missing from this
file. In the future, it probably also makes sense to add some bpf_probe_read_* wrappers
that can read directly into a map pointer, avoiding the BPF stack altogether. I'm going to
call this out of scope for this PR, but plan to submit a subsequent one that addresses
this use case.

alessandrod commented on bpf/aya-bpf/src/helpers.rs:
```diff
@@ -20,6 +40,89 @@ pub unsafe fn bpf_probe_read<T>(src: *const T) -> Result<T, c_long> {
+/// The size of `T` should be the same as the size of whatever is behind `src`. Typically,
+/// `T` would be a binding such as a struct or primitive that directly maps to the pointee
+/// of `src`.
+///
```
The description is valid, but I think it's a bit cryptic. I'd say something like:

```
/// Reads the memory pointed by the _user space_ pointer `src`.
///
/// Returns a bitwise copy of `mem::size_of::<T>()` bytes stored at the user space address `src`. See 
/// `bpf_probe_read_kernel` for  reading kernel space memory.
///
/// # Example
///
/// ...a one liner here...
///
/// # Errors
/// 
/// On failure, returns a negative error code. 
```

alessandrod reviewed (commented):
Thanks a lot for this!

See my comment - I'd rephrase the doc a bit, add examples and put error values in an #Error section for consistency with aya user space.

alessandrod commented:
>  it probably also makes sense to add some bpf_probe_read_* wrappers that can read directly into a map pointer, avoiding the BPF stack altogether. I'm going to call this out of scope for this PR, but plan to submit a subsequent one that addresses this use case.

Yes we should add wrappers that read into a user provided slice, like `bpf_probe_read_user_str` does.

alessandrod commented:
> `bpf_get_current_pid_tgid`

Maybe have `BpfContext::current_pid_tgid()` return a tuple?

willfindlay commented:
@alessandrod I've addressed your comments. Let me know what you think.

Also, I was thinking it might make more sense to just export `gen` instead of `gen::*`. This would help clearly separate our wrappers from the generated helpers and also makes the documentation look _much_ cleaner. WDYT?

alessandrod commented on bpf/aya-bpf/src/helpers.rs:
```diff
@@ -20,6 +54,149 @@ pub unsafe fn bpf_probe_read<T>(src: *const T) -> Result<T, c_long> {
+/// Generally speaking, the more specific [`bpf_probe_read_user_str`] and
+/// [`bpf_probe_read_kernel_str`] should be preferred over this function.
+///
+/// The length of `dest` should be the same or smaller than the length or `src`.
```
This is not true, the destination slice can be larger

alessandrod commented on bpf/aya-bpf/src/helpers.rs:
```diff
@@ -20,6 +54,149 @@ pub unsafe fn bpf_probe_read<T>(src: *const T) -> Result<T, c_long> {
+    Ok(len as usize)
+}
+
+/// Read a null-terminated from _userspace_ stored at `src` into `dest`.
```
missing "string".

Here and in other places, I'd use user space, and kernel space, with a... space :P

alessandrod commented:
🙏🙏🙏
